Content.Client/GameObjects/EntitySystems/BackgroundAudioSystem.cs

```diff
@@ -0,0 +1,181 @@
+
+            _stateManager.OnStateChanged -= StateManagerOnStateChanged;
+
+            _client.PlayerJoinedServer -= OnJoin;
+            _client.PlayerLeaveServer -= OnLeave;
+
+            EndAmbience();
+            EndLobbyMusic();
+        }
+
+        private void StateManagerOnStateChanged(StateChangedEventArgs args)
+        {
+            EndAmbience();
+            EndLobbyMusic();
+            if (args.NewState is LobbyState && _configManager.GetCVar(CCVars.LobbyMusicEnabled))
+            {
+                StartLobbyMusic();
+            }
+            else if (args.NewState is GameScreen && _configManager.GetCVar(CCVars.AmbienceBasicEnabled))
+            {
+                StartAmbience();
+            }
+        }
+
+        private void OnJoin(object? sender, PlayerEventArgs args)
+        {
+            if (_stateManager.CurrentState is LobbyState)
+            {
+                EndAmbience();
+                if (_configManager.GetCVar(CCVars.LobbyMusicEnabled))
+                {
+                    StartLobbyMusic();
+                }
+            }
+            else
+            {
+                EndLobbyMusic();
+                if (_configManager.GetCVar(CCVars.AmbienceBasicEnabled))
+                {
+                    StartAmbience();
+                }
+            }
+        }
+
+        private void OnLeave(object? sender, PlayerEventArgs args)
+        {
+            EndAmbience();
+            EndLobbyMusic();
+        }
+
+        private void AmbienceCVarChanged(bool ambienceEnabled)
+        {
+            if (!ambienceEnabled)
+            {
+                EndAmbience();
+            }
+            else if (_stateManager.CurrentState is GameScreen)
+            {
+                StartAmbience();
+            }
+            else
+            {
+                EndAmbience();
+            }
+        }
+
+        private void StartAmbience()
+        {
+            EndAmbience();
+            var file = _robustRandom.Pick(_ambientCollection.PickFiles);
+            _ambientStream = _audioSystem.Play(file, _ambientParams);
+        }
+
+        private void EndAmbience()
+        {
+            if (_ambientStream == null)
+            {
+                return;
+            }
+            _ambientStream.Stop();
+            _ambientStream = null;
+        }
+
+        private void LobbyMusicCVarChanged(bool musicEnabled)
+        {
+            if (!musicEnabled)
+            {
+                EndLobbyMusic();
+            }
+            else if (_stateManager.CurrentState is LobbyState)
+            {
+                StartLobbyMusic();
+            }
+            else
+            {
+                EndLobbyMusic();
+            }
+        }
+
+        private void StartLobbyMusic()
+        {
+            EndLobbyMusic();
+            var file = _clientGameTicker.LobbySong;
+            if (file == null) // Just in case we fail to get a song from the server, play one at random.
```

Comment:
For some reason we're failing to get the server's song and we're falling back on a randomized one.

Reply:
If I disconnect and then reconnect, all connections after the first work fine.

Reply:
The StateChanged code seems to fire before the client actually receives the net message. Not sure what to do about that.